File: C++/DesignPattern/Singleton.cpp
#include<iostream>
#include<assert.h>
using namespace std;

class Singleton{
private:
  Singleton(){
    cout << "Private Constructor\n";
  }
public:
  static Singleton* getInstance(){
    static Singleton instance;
    return &instance;
  }
};

int main(){
  Singleton* s1 = Singleton::getInstance();
  Singleton* s2 = Singleton::getInstance();
  assert(s1 == s2);
}

File: Java/DesignPatterns/Singleton.java
class World {
    private static World instance = null;

    private World(){
        System.out.println("Private Constructor called");
    }

    public static World getInstance() {
        if(instance == null){instance = new World();}
        return instance;
    }
}

public class Singleton {
    public static void main(String[] args){
        World world1 = World.getInstance();
        World world2 = World.getInstance();
        World world3 = World.getInstance();
    }
}
